### PR TITLE
Dim apply cpu interface and switch cumsum using this interface

### PR DESCRIPTION
**Pull request by glaringlee**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35748 Dim apply cpu interface and switch cumsum using this interface**

Provide a generic dim apply interface that using tensor iterator, support broadcasting, type promotion, etc.

To test this interface, I migrate cumsum cpu version to this interface.
I also moved cumsum code out of cpu folder since cumsum can not benefit from different instruction sets, but will increase the binary size if staying in cpu folder (correct me if I am wrong).

Here is a bench mark compare the previous cumsum implementation and the new dim apply interface.
The benchmark code is the same as https://github.com/pytorch/pytorch/pull/33280.

Previous implementation:
1 thread
size = (10, 10, 10); reduce dim=0; compute time is 0.0119(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0068(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0061(ms)
size = (300, 300, 300); reduce dim=0; compute time is 135.2814(ms)
size = (300, 300, 300); reduce dim=1; compute time is 151.9231(ms)
size = (300, 300, 300); reduce dim=2; compute time is 93.6297(ms)
4 threads
size = (10, 10, 10); reduce dim=0; compute time is 0.0121(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0065(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0060(ms)
size = (300, 300, 300); reduce dim=0; compute time is 55.2960(ms)
size = (300, 300, 300); reduce dim=1; compute time is 61.5722(ms)
size = (300, 300, 300); reduce dim=2; compute time is 40.9760(ms)

New dim apply interface
1 thread
size = (10, 10, 10); reduce dim=0; compute time is 0.0117(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0066(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0063(ms)
size = (300, 300, 300); reduce dim=0; compute time is 130.9699(ms)
size = (300, 300, 300); reduce dim=1; compute time is 148.5702(ms)
size = (300, 300, 300); reduce dim=2; compute time is 91.8762(ms)

4 threads
size = (10, 10, 10); reduce dim=0; compute time is 0.0121(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0063(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0060(ms)
size = (300, 300, 300); reduce dim=0; compute time is 53.8574(ms)
size = (300, 300, 300); reduce dim=1; compute time is 58.0619(ms)
size = (300, 300, 300); reduce dim=2; compute time is 39.7924(ms)

CPU:Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz

Differential Revision: [D20760193](https://our.internmc.facebook.com/intern/diff/D20760193)

### CODE REVIEW

**Comment by glaringlee** [comment]
cc @anjali411 as well

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -188,12 +188,35 @@ static TensorIterator make_reduction(
...
+  if (self.dim() == 0) {
+    result.fill_(self);
+    return result;
+  }
+  if (self.numel() == 0) {
+    return result;
+  }
+
+  TensorIterator iter = dim_apply_op(result, self, maybe_wrap_dim(dim, self.dim()));
+  AT_DISPATCH_ALL_TYPES(iter.dtype(), "cumsum_out_cpu", [&]{
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/ReduceOps.cpp:208]: You should try to move it in kernel code (to compile with AVX) it might give speedup.

**Review by VitalyFedyunin — COMMENTED** [review]
Overall looks good, I need to double-check strides and permutations logic to make sure.

File: aten/src/ATen/native/ReduceOps.cpp (hunk 1 above)
Comment (glaringlee) [reply aten/src/ATen/native/ReduceOps.cpp:208]: No benefit to gain with dim apply case, discussed with Vitaly, we decide to keep this out of cpu folder.

**Comment by glaringlee** [comment]
Close this for now since tensoriterator changed a lot during the past year.

**Pull request closed without merging**